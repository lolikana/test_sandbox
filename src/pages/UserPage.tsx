import { useEffect, useState } from 'react';

import User from '../components/User';

const UserPage = () => {
  const [user, setUser] = useState<{ name: string; email: string } | null>(null);
  const [error, setError] = useState<string>('');
  useEffect(() => {
    fetch('https://jsonplaceholder.typicode.com/users/1')
      .then(response => response.json())
      .then((user: { name: string; email: string }) => setUser(user))
      .catch((error: { message: string }) => setError(error.message));
  }, []);

  if (error) {
    return <span>{error}</span>;
  }

  return (
    <div>
      <h1>User Page</h1>
      {user ? <User name={user.name} email={user.email} /> : <div>Loading...</div>}
    </div>
  );
};

export default UserPage;
