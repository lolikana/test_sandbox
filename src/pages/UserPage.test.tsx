import { render, screen, waitForElementToBeRemoved } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { describe, expect, it, vi } from 'vitest';

import UserPage from './UserPage';

const testResponseData = { name: 'Jack', email: 'jack@test.com' };

const testFetch = vi.fn((url, options) => {
  return new Promise((resolve, reject) => {
    const testResponse = {
      ok: true,
      json() {
        return new Promise((resolve, reject) => {
          resolve(testResponseData);
        });
      }
    };
    resolve(testResponse);
  });
});

vi.stubGlobal('fetch', testFetch);

describe('User Page', () => {
  it('should fint the page title', () => {
    render(
      <MemoryRouter>
        <UserPage />
      </MemoryRouter>
    );
    const title = screen.getByRole('heading', { name: /User Page/i });
    expect(title).toBeInTheDocument();
  });

  it('should show loading text is shown while API request is in progress', async () => {
    render(
      <MemoryRouter>
        <UserPage />
      </MemoryRouter>
    );
    const loading = screen.getByText(/Loading.../i);
    expect(loading).toBeInTheDocument();

    await waitForElementToBeRemoved(() => screen.getByText('Loading...'));
  });

  it("should render user's name", async () => {
    render(
      <MemoryRouter>
        <UserPage />
      </MemoryRouter>
    );

    const userName = await screen.findByText(/Jack/);
    expect(userName).toBeInTheDocument();
  });

  it('should show an error message', async () => {
    vi.spyOn(global, 'fetch').mockRejectedValueOnce({ message: 'API is down' });

    render(
      <MemoryRouter>
        <UserPage />
      </MemoryRouter>
    );
    const errorMessage = await screen.findByText('API is down');
    expect(errorMessage).toBeInTheDocument();
  });
});
