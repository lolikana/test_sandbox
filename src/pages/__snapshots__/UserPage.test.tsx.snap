// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`User Page > should match the snapshot 1`] = `
<DocumentFragment>
  <div>
    <h1>
      User Page
    </h1>
    <div>
      <h2>
        User
      </h2>
      <p>
        Jack
      </p>
      <p>
        jack@test.com
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`User Page > should match the snapshot loading screen 1`] = `
<DocumentFragment>
  <div>
    <h1>
      User Page
    </h1>
    <div>
      Loading...
    </div>
  </div>
</DocumentFragment>
`;
